File: medium/Two-Pointers/reverse-string/reverse_string.cpp
#include <vector>
using namespace std;

class Solution {
public:
    void reverseString(vector<char>& s) {
        /*
         * Reverse char array in-place with two pointers.
         * Approach: Left Right Two Pointers Swap
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        int left = 0;
        int right = s.size() - 1;

        // Swap characters from both sides to center
        while(left < right){
            swap(s[left], s[right]);
            left++;
            right--;
        }
    }
};

File: medium/Two-Pointers/reverse-string/reverse_string.py
from typing import List

class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Reverse input character array in-place.
        Approach: Left & Right Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Left pointer starts at head, right pointer at tail
        left = 0
        right = len(s) - 1

        # Swap characters until two pointers meet
        while left < right:
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1
